### PR TITLE
Use yesno arrays

### PR DESCRIPTION
**Pull request by benadida**
@eventualbuddha I think we need to include the possibility of a `['yes','no']` array, but otherwise I think this looks good. WDYT?

### CODE REVIEW

File: src/v1/index.ts [hunk 1]
```diff
@@ -88,7 +88,7 @@ function encodeBallotVotesInto(
     if (contestVote) {
       if (contest.type === 'yesno') {
         // yesno votes get a single bit
-        bits.writeBoolean(contestVote === 'yes')
+        bits.writeBoolean(contestVote === ['yes'])
```
Comment (eventualbuddha) [inline src/v1/index.ts]: You can't compare arrays like this, as this will always be `false`.

File: src/v0/index.ts [hunk 2]
```diff
@@ -101,18 +101,19 @@ function encodeYesNoVote(
...
-    case 'yes':
-      return '1'
-    default:
-      throw new Error(
-        `cannot encode yesno vote, expected "no" or "yes" but got ${JSON.stringify(
-          contestVote
-        )}`
-      )
+  if (
+    !(contestVote instanceof Array) ||
```
Comment (eventualbuddha) [inline src/v0/index.ts]:
```suggestion
    !Array.isArray(contestVote) ||
```

File: src/v0/index.ts [hunk 3]
```diff
@@ -101,18 +101,19 @@ function encodeYesNoVote(
...
+    (contestVote.length === 1 && !['yes', 'no'].includes(contestVote[0]))
+  ) {
+    throw new Error(
+      `cannot encode yesno vote, expected ['no'] or ['yes'] but got ${JSON.stringify(
+        contestVote
+      )}`
+    )
   }
+
+  return contestVote[0] === 'yes' ? '1' : '0'
```
Comment (eventualbuddha) [inline src/v0/index.ts:117]: This still implicitly encodes `[]` and `['no']` equivalently. Is that intended?

File: src/v1/index.ts [hunk 4]
```diff
@@ -87,8 +88,22 @@ function encodeBallotVotesInto(
 
     if (contestVote) {
       if (contest.type === 'yesno') {
+        const ynVote = contestVote as YesNoVote
+
+        if (!(ynVote instanceof Array)) {
```
Comment (eventualbuddha) [inline src/v1/index.ts]:
```suggestion
        if (!Array.isArray(ynVote)) {
```

File: src/v1/index.ts [hunk 5]
```diff
@@ -87,8 +88,22 @@ function encodeBallotVotesInto(
...
+
+        if (ynVote.length > 1) {
+          throw new Error(
+            `cannot encode a yes/no overvote: ${JSON.stringify(ynVote)}`
+          )
+        }
+
         // yesno votes get a single bit
-        bits.writeBoolean(contestVote === 'yes')
+        bits.writeBoolean(ynVote[0] === 'yes')
```
Comment (eventualbuddha) [inline src/v1/index.ts:106]: Same question as earlier: what about `[]`?

**Review by eventualbuddha — APPROVED** [review]
A question, and some minor fixes.

File: src/v0/index.ts (hunk 3 above)
Comment (benadida) [reply src/v0/index.ts:117]: oof, good point. How should we be encoding `[]`?

Comment (eventualbuddha) [reply src/v0/index.ts:117]: It means "no vote", right? So I'm guessing for v0 it's just an empty string and for v1 it's an absent vote. I'll noodle on it.

**Pull request merged**